modules/module-mssql/src/replication/MSSQLSnapshotQuery.ts

```diff
@@ -0,0 +1,230 @@
+    const stream = request.toReadableStream();
+    if (this.lastKey == null) {
+      request.query(`SELECT TOP(${this.batchSize}) * FROM ${this.table.toQualifiedName()} ORDER BY ${escapedKeyName}`);
+    } else {
+      if (this.key.typeId == null) {
+        throw new Error(`typeId required for primary key ${this.key.name}`);
+      }
+      request
+        .input('lastKey', this.lastKey)
+        .query(
+          `SELECT TOP(${this.batchSize}) * FROM ${this.table.toQualifiedName()} WHERE ${escapedKeyName} > @lastKey ORDER BY ${escapedKeyName}`
+        );
+    }
+
+    // MSSQL only streams one row at a time
+    for await (const row of stream) {
+      this.lastKey = row[this.key.name];
+      yield row;
+    }
+  }
+
+  private deserializeKey(key: Uint8Array) {
+    const decoded = bson.deserialize(key, { useBigInt64: true });
+    const keys = Object.keys(decoded);
+    if (keys.length != 1) {
+      throw new ServiceAssertionError(`Multiple keys found: ${keys.join(', ')}`);
+    }
+    if (keys[0] != this.key.name) {
+      throw new ServiceAssertionError(`Key name mismatch: expected ${this.key.name}, got ${keys[0]}`);
+    }
+
+    return decoded[this.key.name];
+  }
+}
+
+/**
+ * This performs a snapshot query using a list of primary keys.
+ *
+ * This is not used for general snapshots, but is used when we need to re-fetch specific rows
+ * during streaming replication.
+ */
+export class IdSnapshotQuery implements MSSQLSnapshotQuery {
+  static supports(table: SourceTable | MSSQLSourceTable) {
+    // We have the same requirements as BatchedSnapshotQuery.
+    // This is typically only used as a fallback when ChunkedSnapshotQuery
+    // skipped some rows.
+    return BatchedSnapshotQuery.supports(table);
+  }
+
+  public constructor(
+    private readonly transaction: sql.Transaction,
+    private readonly table: MSSQLSourceTable,
+    private readonly keys: PrimaryKeyValue[]
+  ) {}
+
+  public async initialize(): Promise<void> {
+    // No-op
+  }
+
+  public async *next(): AsyncIterableIterator<sql.IColumnMetadata | sql.IRecordSet<any>> {
+    const metadataRequest = this.transaction.request();
+    metadataRequest.stream = true;
+    const metadataPromise = new Promise<sql.IColumnMetadata>((resolve, reject) => {
+      metadataRequest.on('recordset', resolve);
+      metadataRequest.on('error', reject);
+    });
+    metadataRequest.query(`SELECT TOP(0) * FROM ${this.table.toQualifiedName()}`);
+    const columnMetadata: sql.IColumnMetadata = await metadataPromise;
+    yield columnMetadata;
+
+    const keyDefinition = this.table.sourceTable.replicaIdColumns[0];
+    const ids = this.keys.map((record) => record[keyDefinition.name]);
+
+    const request = this.transaction.request();
+    const stream = request.toReadableStream();
+    request
+      .input('ids', ids)
+      .query(`SELECT * FROM ${this.table.toQualifiedName()} WHERE ${escapeIdentifier(keyDefinition.name)} = @ids`);
```

Comment:
This query is likely wrong (equality match on array of ids), but it actually seems this entire "resnapshot" code path is unused. Can remove the entire `IdSnapshotQuery` class, and the `limited` option on `snapshotTableInTx` and `snapshotTable`.